### PR TITLE
Eliminate string constructor for TestPackage

### LINKED ISSUE #889: v4: Create single constructor method/structure for a TestPackage
From memory, there are currently multiple constructors for a test package, which create different package sturctures which the agent must handle. We should reduce this to a single method.

### PR DESCRIPTION
**Pull request by CharliePoole**
Fixes #889 

This fix enables us to move forward without having to support multiple constructors for `TestPackage`, simplifying tests and new feature development. It's not intended as the last word on the interface provided by `TestPackage`, which we will need to rethink somewhat for the design of the final 4.0 release. This is covered in issues #885 and #919.

The old string constructor created a package with a file name but no subpackages. The list constructor created a package with no file name and subpackages. Named packages are only used internally by the engine, so that constructor has now been replaced by the `Named()` modifier. Addition of a `params` overload makes the TestPackage constructor work the same whether one or multiple test files are used.

### CODE REVIEW

**Comment by CharliePoole** [comment]
Although this is only a refactoring toward 4.0, I've asked for review because of substantive changes to the tests. I'll be happy if someone just looks those over and agrees that what I've done is the way to go. In fact, the tests of what runner is used in various circumstances represent what is currently done. If that's not what we want done in some case, we should flag that as a separate issue.

File: src/NUnitEngine/nunit.engine.tests/Services/Fakes/FakeProjectService.cs [hunk 1]
```diff
@@ -27,7 +27,7 @@ void IProjectService.ExpandProjectPackage(TestPackage package)
             if (_projects.ContainsKey(package.Name))
             {
                 foreach (string assembly in _projects[package.Name])
-                    package.AddSubPackage(new TestPackage(assembly));
+                    package.AddSubPackage(new TestPackage().Named(assembly));
```
Comment (ChrisMaddock) [inline src/NUnitEngine/nunit.engine.tests/Services/Fakes/FakeProjectService.cs:30]: All these changes make me think we should have a `package.AddSubPackage(tests.dll)` method.

Actually, now I've written that - what would you think of that as a replacement public API for `TestPackage().Named()`? So project extensions would instead call that method, rather than creating individual packages separately and putting them onto the TestPackage?

We might need it to be `var subpackage = package.AddSubPackage("tests.dll")` to make it easy to set settings on the subpackages...

What do you think?

File: src/NUnitEngine/nunit.engine.core/Services/DomainManager.cs [hunk 2]
```diff
@@ -296,13 +296,18 @@ public static string GetPrivateBinPath(string basePath, IList<TestPackage> packa
...
+
             List<string> dirList = new List<string>();
             StringBuilder sb = new StringBuilder(200);
 
             foreach( string assembly in assemblies )
             {
                 string dir = PathUtils.RelativePath(
-                    Path.GetFullPath(basePath),
+                    basePath,
                     Path.GetDirectoryName( Path.GetFullPath(assembly) ) );
```
Comment (ChrisMaddock) [inline src/NUnitEngine/nunit.engine.core/Services/DomainManager.cs:311]: What's the reason for the changes in this file, sorry? 🙂

**Review by ChrisMaddock — COMMENTED** [review]
Hi Charlie,

The changes look good - but I'm wondering if we can come up with a more explicit API than the `Named()` method - I feel like whether a package is named or not is a consequence of it's place in the structure, rather than key characteristic.

I think I might have stumbled across an alternative while reviewing this - let me know what you think.

File: src/NUnitEngine/nunit.engine.tests/Services/Fakes/FakeProjectService.cs (hunk 1 above)
Comment (ChrisMaddock) [reply src/NUnitEngine/nunit.engine.tests/Services/Fakes/FakeProjectService.cs:30]: One thought - I guess we would lose the ability to create a standalone subpackage. I don't think there's anywhere that needs to do that outside of NUnit's own tests though, is there? I guess it would force us to come up with a nicer solution for testing... 😅

File: src/NUnitEngine/nunit.engine.api/TestPackage.cs [hunk 3]
```diff
@@ -26,39 +26,40 @@ namespace NUnit.Engine
...
-            ID = GetNextID();
-
-            if (filePath != null)
-            {
-                FullName = Path.GetFullPath(filePath);
-                Settings = new Dictionary<string,object>();
-                SubPackages = new List<TestPackage>();
-            }
+            InitializeSubPackages(testFiles);
         }
```
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.api/TestPackage.cs:40]: No, we didn't because it would never have been used for a single string argument - the named constructor would have been used. I added it primarily for ease of use in tests, but then I thought users like ease of use as well. We could have it as the only constructor, but would have to change two or three places where we use the IList constructor. I can go any way you like on this.

File: src/NUnitEngine/nunit.engine.tests/Services/Fakes/FakeProjectService.cs (hunk 1 above)
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.tests/Services/Fakes/FakeProjectService.cs:30]: I had `AddSubPackage(string)` briefly. But the main use for adding subpackages is in project loader extensions and they need to add settings as well. So if you just add the subpackage, then you have to use an indexer to get it back and add settings. It's easier if they can create the subpackage, add any needed settings and then add it to the original package.

File: src/NUnitEngine/nunit.engine.core/Services/DomainManager.cs (hunk 2 above)
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.core/Services/DomainManager.cs:311]: The var `basepath` is already a full path before we get here.

**Comment by CharliePoole** [comment]
I guess the alternative to `Named(string)` was `AddSubPackages(string)`. As I've explained, extension authors need to create full packages, including a name and settings. One option is to make `FullName` settable and I tried that for a while. But some sort of fluent modifier seemed more natural to me.

File: src/NUnitEngine/nunit.engine.tests/Services/Fakes/FakeProjectService.cs (hunk 1 above)
Comment (ChrisMaddock) [reply src/NUnitEngine/nunit.engine.tests/Services/Fakes/FakeProjectService.cs:30]: We could just have AddSubPackage return the subpackage, couldn't we? 

`var subpackage = package.AddSubPackage("tests.dll")`

**Comment by ChrisMaddock** [comment]
> I guess the alternative to Named(string) was AddSubPackages(string). As I've explained, extension authors need to create full packages, including a name and settings. One option is to make FullName settable and I tried that for a while. But some sort of fluent modifier seemed more natural to me.

Interested in what anyone on @nunit/engine-team would think on this one! ☝️ I think we're into preference territory here - and it's a case where mine and your preferences differ, Charlie! 😄

File: src/NUnitEngine/nunit.engine.api/TestPackage.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChrisMaddock) [reply src/NUnitEngine/nunit.engine.api/TestPackage.cs:40]: Personally, I prefer the clean API with a single constructor, over the convenience of both. 

It's not one I feel incredibly strongly on though - if you particularly disagree here. 🙂

File: src/NUnitEngine/nunit.engine.tests/Services/Fakes/FakeProjectService.cs (hunk 1 above)
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.tests/Services/Fakes/FakeProjectService.cs:30]: Good idea. I'll do that!

File: package-tests.cake [hunk 4]
```diff
@@ -293,19 +293,20 @@ public class ZipPackageTester : NetFXPackageTester
...
-            new ExpectedResult("Failed")
-            {
-                Total = 2 * 37,
-                Passed = 2 * 23,
-                Failed = 2 * 5,
-                Warnings = 0,
-                Inconclusive = 2 * 1,
-                Skipped = 2 * 7
-            }));
+        // TODO: Get extensions working - need new releases due to interface change
```
[the remark this answers is not in the document]
Comment (CharliePoole) [reply package-tests.cake:297]: I'll be doing at least the NUnit Project as a new release while we work on this. Don't want to do too much until the API is established.

File: src/NUnitEngine/nunit.engine.api/TestPackage.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.api/TestPackage.cs:40]: I don't have a strong preference in the abstract, but when I tried out each of the two alternatives separately, I was forced to write ugly code in one place or another. Runners and project loaders would usually have some kind of dynamic list and would be forced to construct an array from it. OTOH the 39 uses of the params constructor would have to add a new [] construct to their code.

File: package-tests.cake (hunk 4 above)
[the remark this answers is not in the document]
Comment (CharliePoole) [reply package-tests.cake:297]: Actually, it may only be the NUnit Project loader that has a problem... it uses the old string constructor, which now looks like a params string[] constructor!

**Comment by CharliePoole** [comment]
I've gone with the "AddSubPackage" suggestion in place of "Named".

**Comment by CharliePoole** [comment]
Ready for re-review

**Comment by ChrisMaddock** [comment]
Looks good, thanks!
